tiledb/sm/consolidator/fragment_meta_consolidator.cc

```diff
@@ -0,0 +1,224 @@
+
+  // Include only fragments with footers / separate basic metadata
+  Buffer buff;
+  const auto& tmp_meta = array.fragment_metadata();
+  std::vector<shared_ptr<FragmentMetadata>> meta;
+  for (auto m : tmp_meta) {
+    if (m->format_version() > 2)
+      meta.emplace_back(m);
+  }
+  auto fragment_num = (unsigned)meta.size();
+
+  // Do not consolidate if the number of fragments is not >1
+  if (fragment_num < 2)
+    return array.close();
+
+  // Write number of fragments
+  RETURN_NOT_OK(buff.write(&fragment_num, sizeof(uint32_t)));
+
+  // Compute new URI
+  URI uri;
+  auto& array_dir = array.array_directory();
+  auto first = meta.front()->fragment_uri();
+  auto last = meta.back()->fragment_uri();
+  auto write_version = array.array_schema_latest().write_version();
+  auto&& [st, name] =
+      array_dir.compute_new_fragment_name(first, last, write_version);
+  RETURN_NOT_OK(st);
+
+  auto frag_md_uri = array_dir.get_fragment_metadata_dir(write_version);
+  RETURN_NOT_OK(storage_manager_->vfs()->create_dir(frag_md_uri));
+  uri =
+      URI(frag_md_uri.to_string() + name.value() + constants::meta_file_suffix);
+
+  // Get the consolidated fragment metadata version
+  auto meta_name = uri.remove_trailing_slash().last_path_part();
+  auto pos = meta_name.find_last_of('.');
+  meta_name = (pos == std::string::npos) ? meta_name : meta_name.substr(0, pos);
+  uint32_t meta_version = 0;
+  RETURN_NOT_OK(utils::parse::get_fragment_version(meta_name, &meta_version));
+
+  // Calculate offset of first fragment footer
+  uint64_t offset = sizeof(uint32_t);  // Fragment num
+  for (auto m : meta) {
+    offset += sizeof(uint64_t);  // Name size
+    if (meta_version >= 9) {
+      offset += m->fragment_uri().last_path_part().size();  // Name
+    } else {
+      offset += m->fragment_uri().to_string().size();  // Name
+    }
+    offset += sizeof(uint64_t);  // Offset
+  }
+
+  // Serialize all fragment names and footer offsets into a single buffer
+  for (auto m : meta) {
+    // Write name size and name
+    std::string name;
+    if (meta_version >= 9) {
+      name = m->fragment_uri().last_path_part();
+    } else {
+      name = m->fragment_uri().to_string();
+    }
+    auto name_size = (uint64_t)name.size();
+    RETURN_NOT_OK(buff.write(&name_size, sizeof(uint64_t)));
+    RETURN_NOT_OK(buff.write(name.c_str(), name_size));
+    RETURN_NOT_OK(buff.write(&offset, sizeof(uint64_t)));
+
+    offset += m->footer_size();
+  }
+
+  // Serialize all fragment metadata footers in parallel
+  std::vector<Buffer> buffs(meta.size());
+  auto status = parallel_for(
+      storage_manager_->compute_tp(), 0, buffs.size(), [&](size_t i) {
+        RETURN_NOT_OK(meta[i]->write_footer(&buffs[i]));
+        return Status::Ok();
+      });
+  RETURN_NOT_OK(status);
+
+  // Combine serialized fragment metadata footers into a single buffer
+  for (const auto& b : buffs)
+    RETURN_NOT_OK(buff.write(b.data(), b.size()));
```

Comment:
This is potentially going to realloc every time we write. What is the maximum size of a fragment metadata? It might be worth keeping track of the total size above and pre-emptively resizing the buffer to accomodate, before we write.

Reply:
(Beyond the scope of this PR, but IMO this points to the need for a mechanism to write into a Tile from multiple buffers. Added as a note in SC-16428 for now)

Reply:
This code was just moved but I filed the following:
https://app.shortcut.com/tiledb-inc/story/16435/fragmentmetaconsolidator-consolidate-optimize-fragment-metadata-footer-writes-to-buffer